pkg/apis/cosigned/v1alpha1/clusterimagepolicy_types.go

```diff
@@ -49,9 +36,79 @@ func (*ClusterImagePolicy) GetGroupVersionKind() schema.GroupVersionKind {
 	return SchemeGroupVersion.WithKind("ClusterImagePolicy")
 }
 
-// ClusterImagePolicySpec holds the desired state of the ClusterImagePolicy (from the client).
+// ClusterImagePolicySpec defines a list of images that should be verified
 type ClusterImagePolicySpec struct {
-	// TODO(#1417): Flesh out the specification from the API spec.
+	Images []ImagePattern `json:"images"`
+}
+
+// ImagePattern defines a pattern and its associated authorties
+// If multiple patterns match a particular image, then ALL of
+// those authorities must be satisfied for the image to be admitted.
+type ImagePattern struct {
+	Glob        string      `json:"glob"`
+	Regex       string      `json:"regex"`
+	Authorities []Authority `json:"authorities"`
+}
+
+// The authorities block defines the rules for discovering and
+// validating signatures.  Signatures are
+// cryptographically verified using one of the "key" or "keyless"
+// fields.
+// When multiple authorities are specified, any of them may be used
+// to source the valid signature we are looking for to admit an
+// image.
+
+type Authority struct {
+	Key     KeyRef     `json:"key"`
+	Keyless KeylessRef `json:"keyless"`
+	Sources []Source   `json:"source"`
+	CTLog   TLog       `json:"ctlog"`
+}
+
+// This references a public verification key stored in
+// a secret in the cosign-system namespace.
+type KeyRef struct {
+	SecretRef *v1.SecretReference `json:"secretref"`
+	// Data contains the inline public key
+	Data string `json:"data"`
+	// KMS contains the KMS url of the public key
+	KMS string `json:"kms"`
+}
+
+// Source specifies the location of the signature
+type Source struct {
+	OCI string `json:"oci"`
+}
+
+// TLog specifies the URL to a transparency log that holds
+// the signature and public key information
+type TLog struct {
+	URL string `json:"url"`
+}
+
+// KeylessRef contains location of the validating certificate and the identities
+// against which to verify. KeylessRef will contain either the URL to the verifying
+// certificate, or it will contain the certificate data inline or in a secret.
+type KeylessRef struct {
+	URL        string             `json:"url"`
+	Identities []Identity         `json:"identities"`
+	CAKey      CAKey              `json:"ca-key"`
+	CAKeyRef   v1.SecretReference `json:"ca-keyref"`
+}
+
+// Identity may contain the issue and/or the subject found in the transparency log.
+// Either field supports a pattern glob.
+type Identity struct {
+	Issuer  string `json:"issuer"`
+	Subject string `json:"subject"`
+}
+
+// CAKey contains inline public-key data
+type CAKey struct {
```

Comment:
I'm wondering if we could add SecretRef here instead of name, my understanding was that it was meant to be used for airgaps, where the CA was on the cluster.
Also, if we then add KMS into this struct, it looks identical to KeyRef, so in there we can just inline it. And we can get rid of L96?

Reply:
And now that I think about it, we don't even need to inline it in KeyRef if we think it won't be needing futureproofing. If we are confident (which I'm not :) ), then we could replace L62 with CAKey. I think however it still would make sense to inline into KeyRef, in case there are other things that show up only for KeyRef.

Reply:
Really good point re: SecretRef, putting a change in for that now.

Rather than replacing the KeyRef in Authority with CAKey, we're thinking we can replace CAKey type with KeyRef type.

Reply:
Yeah, or that :) 👍